control-plane/subcommand/install-cni/command.go

```diff
@@ -0,0 +1,373 @@
+		Name:       defaultName,
+		Type:       defaultType,
+		CNIBinDir:  c.flagCNIBinDir,
+		CNINetDir:  c.flagCNINetDir,
+		Multus:     c.flagMultus,
+		Kubeconfig: c.flagKubeconfig,
+		LogLevel:   c.flagLogLevel,
+	}, nil
+}
+
+// newInstallConfig creates the install configuration that is needed for host level operations.
+func (c *Command) newInstallConfig() (*installConfig, error) {
+	return &installConfig{
+		MountedCNIBinDir: "/host" + c.flagCNIBinDir,
+		MountedCNINetDir: "/host" + c.flagCNINetDir,
+		CNIBinSourceDir:  c.flagCNIBinSourceDir,
+	}, nil
+}
+
+// appendCNIConfig appends the consul-cni configuration to the main configuration file.
+func appendCNIConfig(cfg *config.CNIConfig, srcFile, destFile string, logger hclog.Logger) error {
+
+	// Needed to convert the config struct for inserting
+	// Check if file exists
+	if _, err := os.Stat(srcFile); os.IsNotExist(err) {
+		return fmt.Errorf("source cni config file %s does not exist: %v", srcFile, err)
+	}
+	logger.Debug("appendCNIConfig: using files", "srcFile", srcFile, "destFile", destFile)
+	// This section overwrites an existing plugins list entry for istio-cni
+	existingCNIConfig, err := os.ReadFile(srcFile)
+	if err != nil {
+		return err
+	}
+
+	// convert the consul-cni struct into a map
+	var cfgMap map[string]interface{}
+	err = mapstructure.Decode(cfg, &cfgMap)
+	if err != nil {
+		return fmt.Errorf("error loading Consul CNI config: %v", err)
+	}
+
+	// Convert the json config file into a map. The map that is created has 2 parts:
+	// [0] the cni header ()
+	var existingMap map[string]interface{}
+	err = json.Unmarshal(existingCNIConfig, &existingMap)
+	if err != nil {
+		return fmt.Errorf("error unmarshalling existing CNI config: %v", err)
+	}
+
+	// Get the 'plugins' map embedded inside of the exisingMap
+	plugins, ok := existingMap["plugins"].([]interface{})
+	if !ok {
+		return fmt.Errorf("error reading plugin list from CNI config")
+	}
+
+	// Check to see if 'type: consul-cni' already exists and remove it before appending.
+	// This can happen in a CrashLoop and we end up with many entries in the config file
+	logger.Debug("appendCNIConfig: plugins are", "plugins", plugins)
+	for i, p := range plugins {
+		plugin, ok := p.(map[string]interface{})
+		if !ok {
+			return fmt.Errorf("error reading plugin from plugin list")
+		}
+		if plugin["type"] == "consul-cni" {
+			logger.Debug("appendCNIConfig: found existing consul-cni config, removing it")
+			plugins = append(plugins[:i], plugins[i+1:]...)
+			break
+		}
+	}
+
+	// Append the consul-cni map to the already existing plugins
+	existingMap["plugins"] = append(plugins, cfgMap)
+
+	// Marshal into a new json file
+	existingJSON, err := json.MarshalIndent(existingMap, "", "  ")
+	if err != nil {
+		return fmt.Errorf("error marshalling existing CNI config: %v", err)
+	}
+	existingJSON = append(existingJSON, "\n"...)
+
+	// Write the file out
+	err = os.WriteFile(destFile, existingJSON, os.FileMode(0o644))
```

Comment:
To the permissions, no difference. It is a convention I see in several places when setting the rwx permissions in Golang. I think it is done that way because of hexidecimal being `0x`. So, `0o` is octal.  That convention is in some of our HC code and is all over Kubernetes.

Reply:
thanks!